### ISSUE TITLE
Civetweb v1.13 and PDFjs Issue

### ISSUE THREAD

**Issue by kriztioan** [civetweb/civetweb#942]
I'm using civetweb to start a web-server on port 8080. Navigating to http://localhost:8080/index.cgi will show links, when clicked, that display a PDF-file rendered using PDF.js. As of **v1.13**, PDF.js _on occasion_ throws errors that seem to indicate _malformed and/or incomplete data_; subsequently PDF-files are rendered incorrectly. A forced refresh of the page fixes the problem, as it then renders the PDF correctly. Falling back to **v1.12** resolves the issue entirely. Perhaps my reporting is somewhat anecdotal, but systematically trying several versions of PDF.js shows that civitweb at **v1.13** is somehow causing the errors. Unfortunately, I've been unable to pinpoint where things exactly go wrong with civitweb. My initial suspicion would be that perhaps it could be due to a dropped connection? Again, **v1.12** does not have any of the issue described.

**Comment by bel2125** [comment]
> Navigating to http://localhost:8080/index.cgi will show links, when clicked, that display a PDF-file rendered using PDF.js.

Is your index.cgi only providing links, and then the PDF files are just regular files on the disk?
Or does your cgi file directly send the content of the PDF file?

Are you using the same server configuration file? Is keep_alive enabled? Windows or Linux? 
Did you test if this does also happen for HTTPS?

**Comment by kriztioan** [comment]
Thank you for following up. I understand my issue is anecdotal.

The index.cgi file is only providing links, e.g., to http://localhost:8080/script/pdf.js/web/viewer.html?file=%2Farchive/1.pdf.
The PDF-files are regular files on disk.
The CGI-script does _not_ send the PDF-files.
The PDF-files are retrieved via the web-server by PDFjs; e.g., 'GET /archive/108.pdf HTTP/1.1' via a fetch-request.
My server configuration is:

`"document_root", "web"`
`"cgi_pattern", "**.cgi$"`
`index_files", "index.cgi"`
`enable_directory_listing", "no"`
`listening_ports", "8080"`
`error_log_file", "error.log"`
`static_file_max_age", "0"`
`enable_keep_alive", "yes"`
`keep_alive_timeout_ms", "500"`

I have also tried without enabling keep-alive:

`enable_keep_alive", "no"`
`keep_alive_timeout_ms", "0"`,

which made no difference.

I'm running MacOS 10.15.7.
HTTPS is not enabled.

A PDFjs error looks like:

`PDF.js v2.5.207 (build: 0974d6052)`
`Message: Bad (uncompressed) XRef entry: 88R`
`Stack: BaseExceptionClosure@http://localhost:8080/script/pdf.js/build/pdf.js:1302:29`
`@http://localhost:8080/script/pdf.js/build/pdf.js:1305:2`
`__w_pdfjs_require__@http://localhost:8080/script/pdf.js/build/pdf.js:52:30`
`@http://localhost:8080/script/pdf.js/build/pdf.js:388:32`
`__w_pdfjs_require__@http://localhost:7878/script/pdf.js/build/pdf.js:52:30`
`@http://localhost:8080/script/pdf.js/build/pdf.js:322:41`
`__w_pdfjs_require__@http://localhost:7878/script/pdf.js/build/pdf.js:52:30`
`@http://localhost:8080/script/pdf.js/build/pdf.js:116:18`
`@http://localhost:8080/script/pdf.js/build/pdf.js:119:10`
`webpackUniversalModuleDefinition@http://localhost:8080/script/pdf.js/build/pdf.js:31:50`
`@http://localhost:8080/script/pdf.js/build/pdf.js:32:3`

Checking the traffic in Firefox's inspector, I'm not seeing any (obvious) differences when using **v1.12** vs **v1.13**. I do note that PDFjs does request partial content (response code 206) for the PDF-file. Thinking out loud, since the PDF-files can be relatively large (e.g., 2MB), could there perhaps be an issue where the full file hasn't been send over yet (and not closed) when partial content is requested via a different thread, i.e., is two threads accessing the same file a potential issue?

**Comment by bel2125** [comment]
>I do note that PDFjs does request partial content (response code 206) for the PDF-file.

So, the request always has a "Range: " Header?

> Thinking out loud, since the PDF-files can be relatively large (e.g., 2MB), could there perhaps be an issue where the full file hasn't been send over yet (and not closed) when partial content is requested via a different thread, i.e., is two threads accessing the same file a potential issue?

This could only happen, if two connections (two threads) open the same file at the same time.
"Usually" this works, since both connections open the file "read only" - however I have little experience with MacOS. This is the only system I do not have available for testing.
A possible test for this would be to configure `num_threads` as `1` (usually I strongly recommend not to go below 5 to 10).

**Comment by kriztioan** [comment]
First the entire file is request, i.e., _without_ the "Range: " Header. Subsequently, a second request is made for the first 64k of the file, which, I presume, is used to generate the table-of-contents or the like ... I've also seen a subsequent third partial request being made when the file has many pages.

MacOS is POSIX-compliant, so opening the same file for read-only should not really pose a problem, even if a lock would be in place the read should only block ...

Perhaps my situation is some kind of weird edge-case and someone else will encounter an issue that directly points to the culprit here. I will check if setting `num_threads` as `1`  gives some information and otherwise see if I can device some other test.

**Comment by kriztioan** [comment]
I tested the server with setting `num_threads` as `1`. Unfortunately that didn't help. Subsequently I did a `diff` on the raw PDF-data received by the browsers when using **v1.12** and **v1.13**, both for the full file and the "Range:" Header data, but it showed no difference. Considering perhaps a timing issues could be the cause, I set Firefox to simulate a 'Good 3G' connection, however, the PDFjs error remained. Lastly, I tried using a different browser, Safari, and the issue remains. As said, perhaps my situation is some kind of weird edge-case and someone else will encounter an issue that more directly points to the culprit here. For my purposes, remaining at **v1.12** is fine for now.
